server/src/main/java/org/elasticsearch/cluster/coordination/JoinTaskExecutor.java

```diff
@@ -41,19 +41,22 @@
 
     private final AllocationService allocationService;
     private final RerouteService rerouteService;
-    private final long term;
 
-    public JoinTaskExecutor(AllocationService allocationService, RerouteService rerouteService, long term) {
+    public JoinTaskExecutor(AllocationService allocationService, RerouteService rerouteService) {
         this.allocationService = allocationService;
         this.rerouteService = rerouteService;
-        this.term = term;
     }
 
     @Override
     public ClusterState execute(ClusterState currentState, List<TaskContext<JoinTask>> joinTaskContexts) throws Exception {
         // The current state that MasterService uses might have been updated by a (different) master in a higher term already. If so, stop
         // processing the current cluster state update, there's no point in continuing to compute it as it will later be rejected by
         // Coordinator#publish anyhow.
+        assert joinTaskContexts.isEmpty() == false : "Expected to have non empty join tasks list";
+
+        var term = joinTaskContexts.stream().mapToLong(t -> t.getTask().term()).max().getAsLong();
+        joinTaskContexts = joinTaskContexts.stream().filter(t -> t.getTask().term() == term).toList();
```

Comment:
We need to complete every context, but this would not complete the stale ones. Think that means we're missing a unit test that submits tasks with a mix of terms - ideally use something like `ClusterStateTaskExecutorUtils.executeHandlingResults` which verifies that all contexts are completed.

Reply:
I converted this to draft as I wanted to see the complete list of failures. Currently they are:
* CoordinatorTests » testDoesNotElectNonMasterNode
* CoordinatorTests » testNodeCannotJoinIfJoinValidationFailsOnJoiningNode
* CoordinatorTests » testJoiningNodeReceivesFullState
* CoordinatorTests » testDoesNotPerformElectionWhenRestartingFollower
* VotingOnlyNodeCoordinatorTests » testRepeatableTests
* VotingOnlyNodeCoordinatorTests » testDoesNotElectVotingOnlyMasterNode

I also still have not added new test to `JoinTaskExecutorTests`.

Reply:
Updated to throw `JoinTaskOutdatedException` and added the unit test for it. I believe I still need to add it to `ElasticsearchExceptionHandle`